### ISSUE TITLE
NLB support

### ISSUE THREAD

**Issue by carlosonunez** [kubernetes-sigs/cluster-api-provider-aws#3088]
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

As an engineer looking to deploy Kubernetes with CAPA
I would like my control plane to be fronted by a Network Load Balancer instead of Classic ELBs
So that I can continue to use CAPA after AWS deprecates them

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

Classic ELBs are getting deprecated in August 2022, and AWS SDK for Go supports NLBs. We should use them!

More info [here](https://aws.amazon.com/blogs/aws/ec2-classic-is-retiring-heres-how-to-prepare/)

**Comment by sedefsavas** [comment]
Thanks for bringing this up @carlosonunez 
I think we are tracking this somewhere else, will find out and link these. 
There is an ongoing test to see how to deal with hairpinning issue comes with NLBs (when LB directs Kubelet traffic to the api-server on the same node).

**Comment by carlosonunez** [comment]
That's great that this is being tracked elsewhere! I couldn't find any issues pertaining to it and thought "That can't be right!" :)

**Comment by cbusch-pivotal** [comment]
Any update on whether this truly is being tracked? thanks

**Comment by dlipovetsky** [comment]
I haven't seen issue tracking NLB support, but we considered using NLB back in 2018: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/115#discussion_r220562864

**Comment by dlipovetsky** [comment]
> There is an ongoing test to see how to deal with hairpinning issue comes with NLBs (when LB directs Kubelet traffic to the api-server on the same node).

For reference, CAPZ also has a hairpinning issue, although specifically when using an _internal_ load balancer: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1033#issuecomment-1007789608. The current workaround is to prevent requests from kubelet on the control plane machines from going through the LB, by resolving NLB DNS record to 127.0.0.1.

**Comment by sedefsavas** [comment]
I could not find a specific issue for this, I see this was discussed in CAPA v1alpha4 API changes document but during today's office hours, we triaged this issue and keeping this issue.

**Comment by sedefsavas** [comment]
> For reference, CAPZ also has a hairpinning issue, although specifically when using an _internal_ load balancer: [kubernetes-sigs/cluster-api-provider-azure#1033 (comment)](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1033#issuecomment-1007789608). The current workaround is to prevent requests from kubelet on the control plane machines from going through the LB, by resolving NLB DNS record to 127.0.0.1.

Thanks @dlipovetsky, good to know there is an example workaround. If I can't make it work by disabling client IP preservation, this might be the way to go.

**Comment by lubronzhan** [comment]
As suggested by @sedefsavas, changing the target group type from instance to IP, can also workaround this issue https://aws.amazon.com/premiumsupport/knowledge-center/target-connection-fails-load-balancer/
But the downside of this is the IP of instance might change during lifecycle, the corresponding target group needs to be updated, also the NLB's listeners needs to be updated as well. This could cause downtime

**Comment by sedefsavas** [comment]
Just a correction on retirement of Classic load balancers mentioned in the issue:

**EC2 Classic Networking** is being retired not the **Classic Load Balancers**.
Classic Load Balancers that are being used in EC2 Classic networking will need to be migrated to an LB in a VPC. If Classic Load Balancers are already in a VPC, they won't be affected.

TL;DR August 2022 retirement date is not for Classic ELBs.

**Comment by carlosonunez** [comment]
The language is SUPER confusing, but according to this AWS article, ELBs (aka Classic LBs) are in scope for retirement as well: https://docs.aws.amazon.com/elasticloadbalancing/latest/userguide/migrate-classic-load-balancer.html#migrate-step-by-step-classiclink

**Comment by anmolsachan** [comment]
I just wanted to comment, this feature is very much required in the market irrespective of the CLB still being there.

**Comment by vincepri** [comment]
Are we planning to convert the existing Classic ELBs to NLBs?

**Comment by sedefsavas** [comment]
While implementing this, we can avoid introducing the problem of creating LBs with identical names for clusters created with same name in different namespaces: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/969

NLB names need to be unique in the same AWS account and region.

**Comment by Skarlso** [comment]
Hello all.

So I read through this issue finally. :)

What is the consensus here? Do we just disable IP address preservation, or do we do a similar workaround as Azure did?

@sedefsavas do you have anything on this by any chance? It's okay if not, I'm just gathering as much information as I can. :)

**Comment by voor** [comment]
This comment:
> But the downside of this is the IP of instance might change during lifecycle, the corresponding target group needs to be updated, also the NLB's listeners needs to be updated as well. This could cause downtime.
I'm not sure how much of a downside this is, are you implying that when an EC2 instance gets rebooted it might get assigned a new IP Address? That seems like an edge case that might be worth further analysis.  However, when new machines are rolled out they need to be added to the target groups *anyway* -- regardless of ELB or NLB -- and so I don't consider that scenario a "new problem" with NLB specifically.

**Comment by carlosonunez-vmw** [comment]
If the NLB target type for each instance is by instance-id, then this seems like a non-issue?

**Comment by voor** [comment]
> by instance-id,

It looks like Amazon has provided better clarity on this, and if you turn off `Preserve client IP addresses` you can have loopback without using IP Addresses explicitly, this however would mean that the API server would always see the NLB's IP Address even for external users, which does not sound good from an audit standpoint.

Sources: https://docs.aws.amazon.com/elasticloadbalancing/latest/network/load-balancer-troubleshooting.html
Also: https://aws.amazon.com/premiumsupport/knowledge-center/target-connection-fails-load-balancer/

**Comment by Skarlso** [comment]
Yeah, we probably will have to do something like what CAPZ did.

**Comment by Skarlso** [comment]
Okay so last calls. We are going to do something like CAPZ in which they edited the hosts file and added references to the lbs ip as localhost. All in favour say yay all against say nay.

**Comment by carlosonunez** [comment]
yayOn Oct 24, 2022, at 23:49, Gergely Brautigam ***@***.***> wrote:﻿
Okay so last calls. We are going to do something like CAPZ in which they edited the hosts file and added references to the lbs ip as localhost. All in favour say yay all against say nay.

**Comment by johannesfrey** [comment]
> Okay so last calls. We are going to do something like CAPZ in which they edited the hosts file and added references to the lbs ip as localhost. All in favour say yay all against say nay.

Just a naive question. Does this affect the kubelet certificate in any way? E.g. that it also has to contain 127.0.0.1 instead only the node IP?

**Comment by Skarlso** [comment]
>Just a naive question. Does this affect the kubelet certificate in any way? E.g. that it also has to contain 127.0.0.1 instead only the node IP?

Good question. I don't know. :) I will have to consult some documentation. I would assume so; otherwise, the call to localhost might fail with a certificate issue. But also, the call to localhost does not usually require a certificate.

**Comment by Skarlso** [comment]
>Futhermore, would it make sense to make this behavior (map the LB IPs to localhost in /etc/hosts) configurable, as there might be cases where the hairpinning issue is not present and it's desirable that the traffic goes to the API LB instead of the single API server running on the local node.

Certainly. :) That was my intention.

**Comment by Skarlso** [comment]
Yeah, since we will add 127.0.0.1 which is the loopback address, I don't think that is required to be added to the certificate. But let's make sure. :)

**Comment by Skarlso** [comment]
I started to flash this out. I'm going to try and keep it super simple at first. Let's see how that goes.

**Comment by Skarlso** [comment]
Hm, I'm in a bit of a pickle. I was wondering if anyone knows this, perhaps.

I have to register an instance with the NLB. ( I'm not 100% sure how much I have to do this manually, and the NLB doesn't do it. I know I could register an NLB with an ASG, but that's even more complex. )

The problem is that to register an instance, I need a specific TargetGroup. And there could be many target groups with different ports. How the heck do I know in which target group I'm supposed to register the instance? :D

**Comment by Skarlso** [comment]
Eh, for classic we also allow multiple listeners, so there are multiple ports on this thing and each port is a different target group.

And we have no idea what open ports there are on the given instance which we are trying to register.

**Comment by Skarlso** [comment]
This is a sample register call:
```go
	svc := elbv2.New(session.New())
	input := &elbv2.RegisterTargetsInput{
		TargetGroupArn: aws.String("arn:aws:elasticloadbalancing:us-west-2:123456789012:targetgroup/my-new-targets/3bb63f11dfb0faf9"),
		Targets: []*elbv2.TargetDescription{
			{
				Id:   aws.String("i-80c8dd94"),
				Port: aws.Int64(80),
			},
			{
				Id:   aws.String("i-80c8dd94"),
				Port: aws.Int64(766),
			},
		},
	}

	result, err := svc.RegisterTargets(input)
```

The port and targetgroup arn are requires. What if I registered the instance with all of them? :D

**Comment by voor** [comment]
> Like, the api server port is 443.
The API Server port is typically `6443` as well.  You would add all of the control plane instances to a target group mapping 6443 to 6443 (since you are correct, the instances use host networking and therefore are accessible through 6443 on the instance as well).  This is done since the CNI is not yet available on a new k8s cluster, and you would also be in a bit of a chicken-egg situation if you tried to leverage the CNI for the API Server.

**Comment by Skarlso** [comment]
Ah, that's interesting and makes sense. However, that means that users will not be able to specify any custom listeners / target groups they want. Which might be fine?

**Comment by voor** [comment]
> Ah, that's interesting and makes sense. However, that means that users will not be able to specify any custom listeners / target groups they want. Which might be fine?

So you'd determine the port by the Cluster spec, think it's under `spec.controlPlaneEndpoint.port` and that NLB would be effectively used only for the API Server.  So if the end user wanted to use 443, they would set that to be 443, or if they wanted 8080, they'd set it to 8080, etc.

The only thing exposed through these load balancers is the API Server, since etcd pooling is conducted elsewhere, I would consider it out of scope to expose anything other than the API Server through this load balancer.

It might be a nice feature for things like IRSA at a later date, but again would consider that a non-goal in this work.

**Comment by Skarlso** [comment]
Yep, both are right. I'm doing this:

```go
if destinationTG == nil {
		return errors.Errorf("failed to find a target group for AZ '%s'", instanceAZ)
	}

	input := &elbv2.RegisterTargetsInput{
		TargetGroupArn: destinationTG.TargetGroupArn,
		Targets: []*elbv2.TargetDescription{
			{
				AvailabilityZone: aws.String(instanceAZ),
				Id:               aws.String(i.ID),
				Port:             aws.Int64(int64(s.scope.APIServerPort())),
			},
		},
	}

	_, err = s.ELBV2Client.RegisterTargets(input)
	return err
```

which I think is the right way around.. And then the target group is like this:

```go
	Listeners: []infrav1.LBListener{
			{
				Protocol: infrav1.LBProtocolTCP,
				Port:     6443,
				TargetGroup: infrav1.LBTargetGroupSpec{
					Name:     aws.String("apiserver-target"),
					Port:     aws.Int64(6443),
					Protocol: infrav1.LBProtocolTCP,
					VpcID:    aws.String(s.scope.VPC().ID),
					HealthCheck: &infrav1.TargetGroupHealthCheck{
						HealthCheckProtocol:        aws.String(string(infrav1.LBProtocolTCP)),
						HealthCheckIntervalSeconds: aws.Int64(int64(10 * time.Second)),
						HealthCheckTimeoutSeconds:  aws.Int64(int64(5 * time.Second)),
						HealthyThresholdCount:      aws.Int64(5),
					},
				},
			},
		},
	}
```

which might be the wrong way around. :D not a 100% sure. I'll know more when I test it. :)

**Comment by Skarlso** [comment]
All right. Another question for you peeps.

In case of an NLB you can't define Security Groups. So basically, you either leave client ip preservation on and edit the security group so that it allows clients to connect to 6443, or you create a load balancer SG that allows `0.0.0.0/0` or `::/0` to connect to port 6443. This load balancer SG is then attached to the main instance, but that's basically THE SAME THING. :D 6443 will be open.

Other options are firewall rules and the client ip preservation. What do people think about thi?

**Comment by Skarlso** [comment]
As a side note, generally, the consensus on the internet is that you should allow through the SG the client IP to connect to the machine. Which makes sense, I guess. I can determine the client IP of the NLB's DNS record and update the LB SG to allow a connection. That should take care of the Health Check, at the very least.

**Comment by johannesfrey** [comment]
> All right. Another question for you, peeps.
> 
> In the case of an NLB, you can't define Security Groups. So basically, you either leave client IP preservation on and edit the security group so that it allows clients to connect to 6443, or you create a load balancer SG that allows `0.0.0.0/0` or `::/0` to connect to port 6443. This load balancer SG is then attached to the primary instance, but that's basically THE SAME THING. :D 6443 will be open.
> 
> Other options are firewall rules and client IP preservation. What do people think about this?

Most probably I did not understand it correctly. So with NLBs, do you have to set the IP of the control-plane ec2 instance as **target** in the SG or the IP of the load balancer ec2 instance? The `0.0.0.0/0` you are speaking of means the **source** in the SG? Wouldn't that mean that port 6443 is completely exposed?
Do you know what happens if another SG would be created that uses a more specific source (e.g. `/32`). Would the `0.0.0.0/0` always overrule or can you somewhat order their precedence? We certainly have the use case to restrict the access to the api-server lb to a specific IP or range.

**Comment by voor** [comment]
In my experience, the control plane EC2 instance should reside in a private subnet, so the only means that traffic can reach the EC2 instance is through the NLB that resides in a public subnet.  This remediates opening the EC2 instance security group directly to `0.0.0.0/0`.

**Comment by Skarlso** [comment]
@johannesfrey 
>Wouldn't that mean that port 6443 is completely exposed?

Correct. That's why I wouldn't do that.

@voor 
>This remediates opening the EC2 instance security group directly to 0.0.0.0/0.

You still have to open 6443 to internal IPs in the same VPC. Because the NLB doesn't have security groups as the source or assign, it basically acts like any other ec2 or anything that makes requests to an instance. You can't make an NLB a source of something. You can try and detect its IP address and add that specifically, but that doesn't work for various other reasons. 

The only sensible solution I can think of is having the VPC's internal IP address added to one of the security groups as incoming traffic, allowing 6443 as the destination.

And yes, the control plane nodes are usually in a private subnet.

**Comment by Skarlso** [comment]
Uh, that's a good point.

**Comment by Skarlso** [comment]
Hmm, I mean, generally having multiple clusters in the same VPC will end up with that cluster be able to access other clusters on the same network sort of, right?

**Comment by voor** [comment]
Just wondering if you're over-rotating on something that shouldn't be considered a problem -- what is the concern with something going straight to the EC2 instance's port 6443?  It's not being advertised, and even if something did hit it directly it's the same security as if it went through the NLB.  There is nothing magical happening at the NLB to shape or modify that traffic, the NLB is effectively some iptables magic itself to handle load balancing and forwarding the proper traffic.

**Comment by Skarlso** [comment]
@voor Right. However, opening up 6443 to the world is a bit of a stretch, isn't it? :) Even if it isn't advertised. And yes, NLB is just some traffic routing magic it doesn't modify the request at all. But you still would want some amount of security regarding traffic coming to a port. You generally don't open any ports to the outside world.

**Comment by Skarlso** [comment]
BTW, on it being public. With ipv6 there are no private addresses. So this because either a moot. If you use an IPv6 VPC the port will, "technically" be open to the world anyways.

**Comment by Skarlso** [comment]
@voor  I get what you are saying, the reason I need the port opened is that right now it only has it open for a security group as a source. Which doesn't work with the NLB. But yes, since the node is in a private network, opening it to `0.0.0.0/0` _shouldn't_ be a problem, but I would like to be absolutely sure about that. :D 

For example, I'm not sure / don't know if you can have control plane nodes in public networks if you so choose.

**Comment by Skarlso** [comment]
You can technically have control plane nodes in public subnets if you bring your own cluster. I'm saying _technically_ only because it's not really advised... :D

**Comment by Skarlso** [comment]
Successfully created an NLB. Now, lots of testing remaining and making it work with IPv6. ( that shouldn't be a problem. :) )

**Comment by johannesfrey** [comment]
> @voor I get what you are saying, the reason I need the port opened is that right now it only has it open for a security group as a source. Which doesn't work with the NLB. But yes, since the node is in a private network, opening it to `0.0.0.0/0` _shouldn't_ be a problem, but I would like to be absolutely sure about that. :D
> 
> For example, I'm not sure / don't know if you can have control plane nodes in public networks if you so choose.

I'm just wondering (if opening to `0.0.0.0/0` should be the default way to go here) if we still have a (configurable) way to restrict the access to the api-server for specific clients? Or is this a general "problem" of NLBs, as they don't work with SGs and the SG should rather be seen in the conjunction with the actual EC2 instance of the control-plane? If so, in addition to the client IP that should be granted access, to you know if the internal IP of the NLB must also be added to the target security group of the instance?

**Comment by Skarlso** [comment]
The NLB doesn't really have an internal IP OR it's constantly changing. So that would be a difficult manner. However, since the NLB is part of the VPC, it's enough to add the VPC CIDR to the security group. That way, any IP that the NLB will get will be allowed, and at the very least, it's not `0.0.0.0`. :)

And yes, that seems to be the consensus when dealing with NLBs since they don't have security group attachments similar to NAT gateways. It's a Layer 4 load balancer. Historically though, as I've read in many forums by AWS employees, there is no REAL reason that they couldn't make it work with security groups; it's a deliberate decision. Which we have no privy to, sadly. :)

**Comment by johannesfrey** [comment]
Ah, sure. `0.0.0.0` is not necessary :slightly_smiling_face:.
Yeah, really a bummer that NLBs do not work with security groups in the usual way.
But speaking of a SG in front of the EC2 instance of the control-plane, it would not suffice to add the client IP as source (with client IP preservation on) because in reality the NLB connects to that instance (with an unknown internal IP)?

On a side note: @Skarlso, thanks so much for your quick replies and that you are driving this forward. Awesome! :+1:

**Comment by Skarlso** [comment]
No problem. :) I like working on this issue. :D It's very complex. :)

> But speaking of a SG in front of the EC2 instance of the control-plane, it would not suffice to add the client IP as source (with client IP preservation on) because in reality the NLB connects to that instance (with an unknown internal IP)?

I'm not a 100% sure yet to be honest. I don't see with what IP address the nlb tries contacting the instance. If it's using preservation I'm not sure which client ip it preserves.

**Comment by Skarlso** [comment]
Incidentally, that might be the cause of my current problem. Namely that the instance is not joining the cluster.

**Comment by Skarlso** [comment]
BOOM

```
➜  k get nodes --kubeconfig test.kubeconfig
NAME                                           STATUS   ROLES                  AGE     VERSION
ip-10-0-129-98.eu-central-1.compute.internal   Ready    control-plane,master   5m55s   v1.22.1


➜  cluster clusterctl describe cluster test-aws-cluster
NAME                                                                 READY  SEVERITY  REASON  SINCE  MESSAGE
Cluster/test-aws-cluster                                             True                     6m2s
├─ClusterInfrastructure - AWSCluster/test-aws-cluster                True                     8m30s
└─ControlPlane - KubeadmControlPlane/test-aws-cluster-control-plane  True                     6m2s
  └─Machine/test-aws-cluster-control-plane-pjnbs                     True                     8m10s

```

**Comment by Skarlso** [comment]
Okay, so this work in _one_ of two ways:

Either:
- Disable client IP preservation and allow for the VPC's private IP address to go through only
Or:
- Enable client IP preservation, but in that case, I have to open it up with `0.0.0.0/0` for it to work

What I'm going to do is, create a parameter/switch that lets the user control this behaviour. Is that okay?

**Comment by johannesfrey** [comment]
> Okay, so this work in _one_ of two ways:
> 
> Either:
> 
>     * Disable client IP preservation and allow for the VPC's private IP address to go through only
>       Or:
> 
>     * Enable client IP preservation, but in that case, I have to open it up with `0.0.0.0/0` for it to work
> 
> 
> What I'm going to do is, create a parameter/switch that lets the user control this behaviour. Is that okay?

Ah ok. So with client IP preservation on you must open it up for `0.0.0.0/0` because you cannot "know" beforehand which clients tries to connect (e.g. from another VPC or externally)? In contrast to client IP preservation being off, the source will always be the NLB, which resides in the same VPC?!

**Comment by Skarlso** [comment]
>Ah ok. So with client IP preservation on you must open it up for 0.0.0.0/0 because you cannot know beforehand which client tries to connect (e.g. from another VPC or externally)? In contrast to client IP preservation being off, the source will always be the NLB IP, which resides in the same VPC?!

Correct. 👍

**Comment by johannesfrey** [comment]
> > Ah ok. So with client IP preservation on you must open it up for 0.0.0.0/0 because you cannot know beforehand which client tries to connect (e.g. from another VPC or externally)? In contrast to client IP preservation being off, the source will always be the NLB IP, which resides in the same VPC?!
> 
> Correct. +1

Ok, got it. One naive thought: So instead of prematurely opening it up completely for `0.0.0.0/0` here with CAPA, couldn't we delegate the work to or do it similar to e.g. the aws-load-balancer-controller, which picks up the given CIDR from `service.spec.loadBalancerSourceRanges` and creates the necessary SGs/IngressRules only for this CIDR?

Though, this would depend on a Service of type `LoadBalancer` for the api-server... And this would also probably be too late, as the opening up is required during the Cluster creation phase?! Though, kubelet should not rely on it, as it will talk to the local api-server because of the /etc/hosts rewrite?! But most probably I'm mixing up stuff here :sweat_smile:

**Comment by Skarlso** [comment]
>And this would also probably be too late, as the opening up is required during the Cluster creation phase

This is correct. Without opening it up the node will not join the cluster and kubelet will fail with various errors or things that it can't reach.

Some of which MIGHT be the hair pinning issue which I'm investigating. For example just this:

```
Nov 04 08:23:45 ip-10-0-118-234 kubelet[1810]: E1104 08:23:45.087874    1810 kubelet.go:2332] "Container runtime network not ready" networkReady="NetworkReady=false reason:NetworkPluginNotReady message:Network plugin returns error: cni plugin not initialized"
Nov 04 08:23:45 ip-10-0-118-234 kubelet[1810]: E1104 08:23:45.156058    1810 kubelet.go:2407] "Error getting node" err="node \"ip-10-0-118-234.eu-central-1.compute.internal\" not found"
Nov 04 08:23:45 ip-10-0-118-234 kubelet[1810]: E1104 08:23:45.256934    1810 kubelet.go:2407] "Error getting node" err="node \"ip-10-0-118-234.eu-central-1.compute.internal\" not found"
``` 

Looks very suspiciously like a hair pinning issue where kubelet can't access the node using its domain name.

**Comment by Skarlso** [comment]
Okay, so I got something working for now. I'm going to roll with that, then iterate over the possible hair pinning issue.

**Comment by Skarlso** [comment]
So I tried the `hosts` rewrite. Sadly, it didn't work. :(

```
Nov 06 06:22:43 ip-10-0-120-185 kubelet[2114]: E1106 06:22:43.455629    2114 kubelet.go:2407] "Error getting node" err="node \"ip-10-0-120-185.eu-central-1.compute.internal\" not found"
Nov 06 06:22:43 ip-10-0-120-185 kubelet[2114]: E1106 06:22:43.554024    2114 controller.go:144] failed to ensure lease exists, will retry in 7s, error: Get "https://ypcqehss1nfq26i6gfooheufe2ok-k8s-3f5da43546a9b6c8.elb.eu-central-1.amazonaws.com:6443/apis/coordination.k8s.io/v1/namespaces/kube-node-lease/leases/ip-10-0-120-185.eu-central-1.compute.internal?timeout=10s": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Nov 06 06:22:43 ip-10-0-120-185 kubelet[2114]: E1106 06:22:43.558110    2114 kubelet.go:2407] "Error getting node" err="node \"ip-10-0-120-185.eu-central-1.compute.internal\" not found"

```

So for now, I'm staying with my solution. FFW the network loadbalancer doc troubleshoot section also says that we should disable preserve client ip if hairpinning is an issue.

**Comment by voor** [comment]
Does it set the X-Forwarded-For request header or X-Real-Ip header?  That way audit logs will still have the client IP somewhere.

**Comment by Skarlso** [comment]
It does, yes.

**Comment by Skarlso** [comment]
Ah, yeah, I misread the doc. You're right @voor it does not set that header. Something else does if client ip preservation is on.

**Comment by johannesfrey** [comment]
Just out of interest, as it's seems not be tracked in a milestone. I assume it won't be part of the [v2.0.0](https://github.com/kubernetes-sigs/cluster-api-provider-aws/milestone/36) milestone?!

**Comment by johannesfrey** [comment]
> Probably not, because the PR needs reviews, tests, etc. That won't happen now, the team is pushing for 2.0.0 milestones and fixes now. Sorry! :/

Sure, no worries :slightly_smiling_face:. Totally undestand. Just noticed that 2.0.0 is about to be released soon, so I thought I ask. Really nice that 2.0.0 is coming, even though without NLB :slightly_smiling_face:

**Comment by Skarlso** [comment]
There are two hanging things:
- e2e test implementation #3820 
- adding SSL for listeners / target group #3899 

But I could consider this closed. Please test and see if anything is missing / not working. :)
